File: BLAS/SRC/ztrsv.f
      SUBROUTINE ZTRSV(UPLO,TRANS,DIAG,N,A,LDA,X,INCX)

*  -- Reference BLAS level2 routine --
*  -- Reference BLAS is a software package provided by Univ. of Tennessee,    --
*  -- Univ. of California Berkeley, Univ. of Colorado Denver and NAG Ltd..--

      // .. Scalar Arguments ..
      int     INCX,LDA,N;
      String    DIAG,TRANS,UPLO;
      // ..
      // .. Array Arguments ..
      COMPLEX*16 A(LDA,*),X(*)
      // ..

*  =====================================================================

      // .. Parameters ..
      COMPLEX*16 ZERO
      const     ZERO= (0.0D+0,0.0D+0);
      // ..
      // .. Local Scalars ..
      COMPLEX*16 TEMP
      int     I,INFO,IX,J,JX,KX;
      bool    NOCONJ,NOUNIT;
      // ..
      // .. External Functions ..
      bool    LSAME;
      // EXTERNAL LSAME
      // ..
      // .. External Subroutines ..
      // EXTERNAL XERBLA
      // ..
      // .. Intrinsic Functions ..
      // INTRINSIC DCONJG,MAX
      // ..

      // Test the input parameters.

      INFO = 0
      if (.NOT.LSAME(UPLO,'U') .AND. .NOT.LSAME(UPLO,'L')) {
          INFO = 1
      } else if (.NOT.LSAME(TRANS,'N') .AND. .NOT.LSAME(TRANS,'T') .AND. .NOT.LSAME(TRANS,'C')) {
          INFO = 2
      } else if (.NOT.LSAME(DIAG,'U') .AND. .NOT.LSAME(DIAG,'N')) {
          INFO = 3
      } else if (N.LT.0) {
          INFO = 4
      } else if (LDA.LT.MAX(1,N)) {
          INFO = 6
      } else if (INCX.EQ.0) {
          INFO = 8
      }
      if (INFO.NE.0) {
          CALL XERBLA('ZTRSV ',INFO)
          RETURN
      }

      // Quick return if possible.

      IF (N.EQ.0) RETURN

      NOCONJ = LSAME(TRANS,'T')
      NOUNIT = LSAME(DIAG,'N')

      // Set up the start point in X if the increment is not unity. This
      // will be  ( N - 1 )*INCX  too small for descending loops.

      if (INCX.LE.0) {
          KX = 1 - (N-1)*INCX
      } else if (INCX.NE.1) {
          KX = 1
      }

      // Start the operations. In this version the elements of A are
      // accessed sequentially with one pass through A.

      if (LSAME(TRANS,'N')) {

         // Form  x := inv( A )*x.

          if (LSAME(UPLO,'U')) {
              if (INCX.EQ.1) {
                  DO 20 J = N,1,-1
                      if (X(J).NE.ZERO) {
                          IF (NOUNIT) X(J) = X(J)/A(J,J)
                          TEMP = X(J)
                          DO 10 I = J - 1,1,-1
                              X(I) = X(I) - TEMP*A(I,J)
   10                     CONTINUE
                      }
   20             CONTINUE
              } else {
                  JX = KX + (N-1)*INCX
                  DO 40 J = N,1,-1
                      if (X(JX).NE.ZERO) {
                          IF (NOUNIT) X(JX) = X(JX)/A(J,J)
                          TEMP = X(JX)
                          IX = JX
                          DO 30 I = J - 1,1,-1
                              IX = IX - INCX
                              X(IX) = X(IX) - TEMP*A(I,J)
   30                     CONTINUE
                      }
                      JX = JX - INCX
   40             CONTINUE
              }
          } else {
              if (INCX.EQ.1) {
                  DO 60 J = 1,N
                      if (X(J).NE.ZERO) {
                          IF (NOUNIT) X(J) = X(J)/A(J,J)
                          TEMP = X(J)
                          DO 50 I = J + 1,N
                              X(I) = X(I) - TEMP*A(I,J)
   50                     CONTINUE
                      }
   60             CONTINUE
              } else {
                  JX = KX
                  DO 80 J = 1,N
                      if (X(JX).NE.ZERO) {
                          IF (NOUNIT) X(JX) = X(JX)/A(J,J)
                          TEMP = X(JX)
                          IX = JX
                          DO 70 I = J + 1,N
                              IX = IX + INCX
                              X(IX) = X(IX) - TEMP*A(I,J)
   70                     CONTINUE
                      }
                      JX = JX + INCX
   80             CONTINUE
              }
          }
      } else {

         // Form  x := inv( A**T )*x  or  x := inv( A**H )*x.

          if (LSAME(UPLO,'U')) {
              if (INCX.EQ.1) {
                  DO 110 J = 1,N
                      TEMP = X(J)
                      if (NOCONJ) {
                          DO 90 I = 1,J - 1
                              TEMP = TEMP - A(I,J)*X(I)
   90                     CONTINUE
                          IF (NOUNIT) TEMP = TEMP/A(J,J)
                      } else {
                          DO 100 I = 1,J - 1
                              TEMP = TEMP - DCONJG(A(I,J))*X(I)
  100                     CONTINUE
                          IF (NOUNIT) TEMP = TEMP/DCONJG(A(J,J))
                      }
                      X(J) = TEMP
  110             CONTINUE
              } else {
                  JX = KX
                  DO 140 J = 1,N
                      IX = KX
                      TEMP = X(JX)
                      if (NOCONJ) {
                          DO 120 I = 1,J - 1
                              TEMP = TEMP - A(I,J)*X(IX)
                              IX = IX + INCX
  120                     CONTINUE
                          IF (NOUNIT) TEMP = TEMP/A(J,J)
                      } else {
                          DO 130 I = 1,J - 1
                              TEMP = TEMP - DCONJG(A(I,J))*X(IX)
                              IX = IX + INCX
  130                     CONTINUE
                          IF (NOUNIT) TEMP = TEMP/DCONJG(A(J,J))
                      }
                      X(JX) = TEMP
                      JX = JX + INCX
  140             CONTINUE
              }
          } else {
              if (INCX.EQ.1) {
                  DO 170 J = N,1,-1
                      TEMP = X(J)
                      if (NOCONJ) {
                          DO 150 I = N,J + 1,-1
                              TEMP = TEMP - A(I,J)*X(I)
  150                     CONTINUE
                          IF (NOUNIT) TEMP = TEMP/A(J,J)
                      } else {
                          DO 160 I = N,J + 1,-1
                              TEMP = TEMP - DCONJG(A(I,J))*X(I)
  160                     CONTINUE
                          IF (NOUNIT) TEMP = TEMP/DCONJG(A(J,J))
                      }
                      X(J) = TEMP
  170             CONTINUE
              } else {
                  KX = KX + (N-1)*INCX
                  JX = KX
                  DO 200 J = N,1,-1
                      IX = KX
                      TEMP = X(JX)
                      if (NOCONJ) {
                          DO 180 I = N,J + 1,-1
                              TEMP = TEMP - A(I,J)*X(IX)
                              IX = IX - INCX
  180                     CONTINUE
                          IF (NOUNIT) TEMP = TEMP/A(J,J)
                      } else {
                          DO 190 I = N,J + 1,-1
                              TEMP = TEMP - DCONJG(A(I,J))*X(IX)
                              IX = IX - INCX
  190                     CONTINUE
                          IF (NOUNIT) TEMP = TEMP/DCONJG(A(J,J))
                      }
                      X(JX) = TEMP
                      JX = JX - INCX
  200             CONTINUE
              }
          }
      }

      RETURN

      // End of ZTRSV

      }

File: lib/src/blas/ztrsv.dart
// Copyright (c) 2024 Guilherme Lepsch. All rights reserved. Use of this
// source code is governed by a BSD-style license that can be found in the
// [LICENSE file](https://github.com/lepsch/dart_lapack/blob/main/LICENSE).

import 'dart:math';

import 'package:lapack/src/blas/lsame.dart';
import 'package:lapack/src/blas/xerbla.dart';
import 'package:lapack/src/complex.dart';
import 'package:lapack/src/matrix.dart';

void ztrsv(
  final String UPLO,
  final String TRANS,
  final String DIAG,
  final int N,
  final Matrix<Complex> A_,
  final int LDA,
  final Array<Complex> X_,
  final int INCX,
) {
// -- Reference BLAS level2 routine --
// -- Reference BLAS is a software package provided by Univ. of Tennessee,    --
// -- Univ. of California Berkeley, Univ. of Colorado Denver and NAG Ltd..--
  final A = A_.having(ld: LDA);
  final X = X_.having();
  Complex TEMP;
  int I, INFO, IX, J, JX, KX = 0;
  bool NOCONJ, NOUNIT;

  // Test the input parameters.

  INFO = 0;
  if (!lsame(UPLO, 'U') && !lsame(UPLO, 'L')) {
    INFO = 1;
  } else if (!lsame(TRANS, 'N') && !lsame(TRANS, 'T') && !lsame(TRANS, 'C')) {
    INFO = 2;
  } else if (!lsame(DIAG, 'U') && !lsame(DIAG, 'N')) {
    INFO = 3;
  } else if (N < 0) {
    INFO = 4;
  } else if (LDA < max(1, N)) {
    INFO = 6;
  } else if (INCX == 0) {
    INFO = 8;
  }
  if (INFO != 0) {
    xerbla('ZTRSV', INFO);
    return;
  }

  // Quick return if possible.

  if (N == 0) return;

  NOCONJ = lsame(TRANS, 'T');
  NOUNIT = lsame(DIAG, 'N');

  // Set up the start point in X if the increment is not unity. This
  // will be  ( N - 1 )*INCX  too small for descending loops.

  if (INCX <= 0) {
    KX = 1 - (N - 1) * INCX;
  } else if (INCX != 1) {
    KX = 1;
  }

  // Start the operations. In this version the elements of A are
  // accessed sequentially with one pass through A.

  if (lsame(TRANS, 'N')) {
    // Form  x := inv( A )*x.

    if (lsame(UPLO, 'U')) {
      if (INCX == 1) {
        for (J = N; J >= 1; J--) {
          if (X[J] != Complex.zero) {
            if (NOUNIT) X[J] /= A[J][J];
            TEMP = X[J];
            for (I = J - 1; I >= 1; I--) {
              X[I] -= TEMP * A[I][J];
            }
          }
        }
      } else {
        JX = KX + (N - 1) * INCX;
        for (J = N; J >= 1; J--) {
          if (X[JX] != Complex.zero) {
            if (NOUNIT) X[JX] /= A[J][J];
            TEMP = X[JX];
            IX = JX;
            for (I = J - 1; I >= 1; I--) {
              IX -= INCX;
              X[IX] -= TEMP * A[I][J];
            }
          }
          JX -= INCX;
        }
      }
    } else {
      if (INCX == 1) {
        for (J = 1; J <= N; J++) {
          if (X[J] != Complex.zero) {
            if (NOUNIT) X[J] /= A[J][J];
            TEMP = X[J];
            for (I = J + 1; I <= N; I++) {
              X[I] -= TEMP * A[I][J];
            }
          }
        }
      } else {
        JX = KX;
        for (J = 1; J <= N; J++) {
          if (X[JX] != Complex.zero) {
            if (NOUNIT) X[JX] /= A[J][J];
            TEMP = X[JX];
            IX = JX;
            for (I = J + 1; I <= N; I++) {
              IX += INCX;
              X[IX] -= TEMP * A[I][J];
            }
          }
          JX += INCX;
        }
      }
    }
  } else {
    // Form  x := inv( A**T )*x  or  x := inv( A**H )*x.

    if (lsame(UPLO, 'U')) {
      if (INCX == 1) {
        for (J = 1; J <= N; J++) {
          TEMP = X[J];
          if (NOCONJ) {
            for (I = 1; I <= J - 1; I++) {
              TEMP -= A[I][J] * X[I];
            }
            if (NOUNIT) TEMP /= A[J][J];
          } else {
            for (I = 1; I <= J - 1; I++) {
              TEMP -= A[I][J].conjugate() * X[I];
            }
            if (NOUNIT) TEMP /= A[J][J].conjugate();
          }
          X[J] = TEMP;
        }
      } else {
        JX = KX;
        for (J = 1; J <= N; J++) {
          IX = KX;
          TEMP = X[JX];
          if (NOCONJ) {
            for (I = 1; I <= J - 1; I++) {
              TEMP -= A[I][J] * X[IX];
              IX += INCX;
            }
            if (NOUNIT) TEMP /= A[J][J];
          } else {
            for (I = 1; I <= J - 1; I++) {
              TEMP -= A[I][J].conjugate() * X[IX];
              IX += INCX;
            }
            if (NOUNIT) TEMP /= A[J][J].conjugate();
          }
          X[JX] = TEMP;
          JX += INCX;
        }
      }
    } else {
      if (INCX == 1) {
        for (J = N; J >= 1; J--) {
          TEMP = X[J];
          if (NOCONJ) {
            for (I = N; I >= J + 1; I--) {
              TEMP -= A[I][J] * X[I];
            }
            if (NOUNIT) TEMP /= A[J][J];
          } else {
            for (I = N; I >= J + 1; I--) {
              TEMP -= A[I][J].conjugate() * X[I];
            }
            if (NOUNIT) TEMP /= A[J][J].conjugate();
          }
          X[J] = TEMP;
        }
      } else {
        KX += (N - 1) * INCX;
        JX = KX;
        for (J = N; J >= 1; J--) {
          IX = KX;
          TEMP = X[JX];
          if (NOCONJ) {
            for (I = N; I >= J + 1; I--) {
              TEMP -= A[I][J] * X[IX];
              IX -= INCX;
            }
            if (NOUNIT) TEMP /= A[J][J];
          } else {
            for (I = N; I >= J + 1; I--) {
              TEMP -= A[I][J].conjugate() * X[IX];
              IX -= INCX;
            }
            if (NOUNIT) TEMP /= A[J][J].conjugate();
          }
          X[JX] = TEMP;
          JX -= INCX;
        }
      }
    }
  }
}
